### PR TITLE
GitHub 7865 haspropwithvaluequery leaks

### LINKED ISSUE #7865: HasPropWithValueQueryBase::getPair() may leak the value of RDValue.
`HasPropWithValueQueryBase::getPair()` returns a `Dict::Pair()`, which contains a `RDValue` member. `RDValue` holds non-POD data in the form of pointers, and since neither `RDValue` nor `Dict::Pair()` have destructors, the values will leak if not cleaned up.

Found by running `pickleTestsCatch  "pickling HasPropWithValue queries"` under asan. It's not trivially fixable, so I'm raising it here instead of rushing a fix.

### PR DESCRIPTION
**Pull request by bp-kelley**
Fixes #7865

### CODE REVIEW

File: Code/GraphMol/MolPickler.cpp [hunk 1]
```diff
@@ -473,6 +473,7 @@ void pickleQuery(std::ostream &ss, const Query<int, T const *, true> *query) {
         streamWrite(ss, MolPickler::QUERY_VALUE, std::get<2>(v));
         streamWriteProp(ss, std::get<1>(v),
                         MolPickler::getCustomPropHandlers());
+        std::get<1>(v).cleanup();
```
Comment (bp-kelley) [inline Code/GraphMol/MolPickler.cpp:476]: This is the "real" fix here.

**Comment by bp-kelley** [comment]
I agree that Dict::Pair was probably the wrong choice here as it really shouldn't be used out of a dictionary.  I'll make a thin wrapper that cleans it up migrate external usage to that and then run a valgrind test for leaks.

**Comment by bp-kelley** [comment]
I'll fix the swig builds this evening.  Need to ignore the PairHolder funcs.

**Comment by ricrogz** [comment]
> This didn't show any leaks in valgrind at least.

asan also doesn't report any mem errors.

**Comment by bp-kelley** [comment]
@greglandrum this one is finally ready as well.

**Comment by greglandrum** [comment]
@bp-kelley since this changes the public API, there should be an entry in the "Backwards Incompatible Changes" part of the release notes.

File: Code/GraphMol/Substruct/catch_tests.cpp [hunk 2]
```diff
@@ -587,14 +589,18 @@ TEST_CASE("pickling HasPropWithValue queries") {
...
     {
       std::string pkl;
-      MolPickler::pickleMol(mol2, pkl);
+      MolPickler::pickleMol(mol3, pkl);
       RWMol pklmol(pkl);
       REQUIRE(pklmol.getAtomWithIdx(0)->hasQuery());
       REQUIRE(pklmol.getAtomWithIdx(0)->hasQuery());
       CHECK(SubstructMatch(*target, pklmol, ps).size() == 0);
+      // make sure we are idempotent in pickling
+      CHECK(SubstructMatch(*target, mol3, ps).size() == 0);
```
Comment (greglandrum) [inline Code/GraphMol/Substruct/catch_tests.cpp:603]: What are you testing with these additions?

File: Code/GraphMol/MolPickler.cpp [hunk 3]
```diff
@@ -465,14 +465,15 @@ void pickleQuery(std::ostream &ss, const Query<int, T const *, true> *query) {
         streamWrite(ss, MolPickler::QUERY_VALUE, pval);
       } break;
       case 6: {
-        auto v = boost::get<std::tuple<MolPickler::Tags, Dict::Pair, double>>(
+        auto v = boost::get<std::tuple<MolPickler::Tags, PairHolder, double>>(
```
Comment (greglandrum) [inline Code/GraphMol/MolPickler.cpp:468]: this is copying the tuple. I think you want a reference.

File: Code/GraphMol/MolPickler.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/MolPickler.cpp:476]: I think it's actually redundant since the destructor of the `PairHolder` calls `cleanup_rdvalue()`.

**Review by greglandrum — CHANGES_REQUESTED** [review]
a couple suggestions here, plus a few that I made accidentally outside the review itself

File: Code/GraphMol/Substruct/catch_tests.cpp (hunk 2 above)
Comment (bp-kelley) [reply Code/GraphMol/Substruct/catch_tests.cpp:603]: This is ensuring that pickling doesn't change the original molecule.  I forget exactly why I added it, but I think I had seen a case where pickling changed the results when the original molecule was used again.

Comment (greglandrum) [reply Code/GraphMol/Substruct/catch_tests.cpp:603]: The only way pickling would change the substructure matching behavior in this case is if the properties change, and that's a test for the pickling suite, not here.

Comment (bp-kelley) [reply Code/GraphMol/Substruct/catch_tests.cpp:603]: This is where we had added the pickle tests before, I'll move them.

File: Code/GraphMol/MolPickler.cpp (hunk 3 above)
Comment (bp-kelley) [reply Code/GraphMol/MolPickler.cpp:468]: Good catch

File: Code/GraphMol/Substruct/catch_tests.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/Substruct/catch_tests.cpp:603]: > This is where we had added the pickle tests before, I'll move them.

There's no need to move anything, it's not like extra tests are bad, but I don't think these are actually testing anything real

Comment (bp-kelley) [reply Code/GraphMol/Substruct/catch_tests.cpp:603]: With the change to PairHolder this shouldn't happen anymore.

**Comment by bp-kelley** [comment]
@greglandrum I think I addressed all the comments.

File: ReleaseNotes.md [hunk 4]
```diff
@@ -11,6 +11,10 @@ GitHub)
 
 ## Backwards incompatible changes
 
+HasPropWithValueQueryBase used RDKit::Dict::Pair to return
```
Comment (greglandrum) [inline ReleaseNotes.md:14]: This should be a bulleted list, please start the first line with `- ` like is in the other sections of this document

**Review by greglandrum — APPROVED** [review]
LGTM

**Pull request merged**